_drafts/code-properties-and-language-features.md

```diff
@@ -0,0 +1,50 @@
+I feel like I only just started learning good object-oriented design, and the world has moved on to functional programming! As a result, the OO principles that are fresh on my mind are exactly what people are moving away from. For example, as I've been checking out Elixir and Redux, I've had questions about whether a lack of certain OO features causes problems. For example, encapsulation allows you to isolate implementation knowledge to one part of your app, so that other parts of your app can use that functionality without being coupled to its implementation. Is it a problem that FP doesn't offer the same kind of encapsulation?
+
+It's difficult for me to find answers to these questions. As Yehuda Katz [tweeted][yehuda-1] [recently][yehuda-2], there's a tendency to describe programming advancements as radically new paradigms, and past approaches as terrible. There's no grappling with the arguments older approaches made.
+
+This leaves me with a conundrum: I bought into arguments about encapsulation increasing maintainability. Were those arguments true, or were they just something we made up to tell ourselves so we could build software in a way we liked?
+
+Even though these things aren't discussed much publicly, asking questions of people who know Elixir and Redux helped me with this.
+
+# Questioning Elixir and Redux
+
+First, I asked about the way data is separate from functionality in Elixir, and whether that caused problems. But I learned that when you define a struct inside a module, you also usually define functions that operate on that struct inside the same module. Because the name of the module and the name of the struct are the same, they go together naturally. So data and functionality aren't entirely separated.
+
+Then I asked about encapsulation: the data in a struct is available for everyone else in the app to access directly. But I learned that often you might have a convention of only accessing data in the struct via a function, and that that is a way to isolate knowledge of the data structure.
+
+Then I asked about [the `Plug.Conn` struct][conn] in the Phoenix framework, which is passed through much of the app. I asked about whether this made the entire application fragile, because all of it required intimate knowledge of the Connection. I learned that there was a [`private`][conn-private] field on the struct, and although you aren't prevented from accessing it directly, the convention is that you shouldn't because the implementation in it can change without warning. This reminded me of some Rubyists' point that, because you can get around the private keyword, it's effectively just a convention as well.
```

Comment:
I'd disagree with the statement that the whole app requires knowledge of `conn`. Ultimately, the only pieces of the app that need to know about `conn` are the ones need to alter the request or the response. I'd argue that for non-trivial apps, the vast majority of the app is in the domain model, which doesn't need to know anything about HTTP at all. It's easy to yank only the things you need out of `conn` (such as request parameters) and pass just those around.